tracer/test/Datadog.Trace.Tests/Telemetry/Collectors/IntegrationTelemetryCollectorTests.cs

```diff
@@ -0,0 +1,182 @@
+            collector.RecordTracerSettings(new ImmutableTracerSettings(new TracerSettings()));
+
+            collector.GetData();
+            collector.HasChanges().Should().BeFalse();
+
+            collector.IntegrationGeneratedSpan(IntegrationId);
+            collector.HasChanges().Should().BeTrue();
+            collector.GetData();
+
+            collector.IntegrationGeneratedSpan(IntegrationId);
+            collector.HasChanges().Should().BeFalse();
+        }
+
+        [Fact]
+        public void HasChangesWhenNewIntegrationDisabled()
+        {
+            var collector = new IntegrationTelemetryCollector();
+            collector.RecordTracerSettings(new ImmutableTracerSettings(new TracerSettings()));
+
+            collector.GetData();
+            collector.HasChanges().Should().BeFalse();
+
+            collector.IntegrationDisabledDueToError(IntegrationId, "Testing!");
+            collector.HasChanges().Should().BeTrue();
+        }
+
+        [Fact]
+        public void DoesNotHaveChangesWhenSameIntegrationDisabled()
+        {
+            var collector = new IntegrationTelemetryCollector();
+            collector.RecordTracerSettings(new ImmutableTracerSettings(new TracerSettings()));
+
+            collector.GetData();
+            collector.HasChanges().Should().BeFalse();
+
+            collector.IntegrationDisabledDueToError(IntegrationId, "Testing");
+            collector.HasChanges().Should().BeTrue();
+            collector.GetData();
+
+            collector.IntegrationDisabledDueToError(IntegrationId, "Another error");
+            collector.HasChanges().Should().BeFalse();
+        }
+
+        [Fact]
+        public void WhenIntegrationRunsSuccessfullyHasExpectedValues()
+        {
+            var collector = new IntegrationTelemetryCollector();
+            collector.RecordTracerSettings(new ImmutableTracerSettings(new TracerSettings()));
+
+            collector.IntegrationRunning(IntegrationId);
+            collector.IntegrationGeneratedSpan(IntegrationId);
+
+            var data = collector.GetData();
+            var integration = data.FirstOrDefault(x => x.Name == IntegrationName);
+            integration.Should().NotBeNull();
+            integration.AutoEnabled.Should().BeTrue();
+            integration.Enabled.Should().BeTrue();
+            integration.Error.Should().BeNull();
+        }
+
+        [Fact]
+        public void WhenIntegrationRunsButDoesNotGenerateSpanHasExpectedValues()
+        {
+            var collector = new IntegrationTelemetryCollector();
+            collector.RecordTracerSettings(new ImmutableTracerSettings(new TracerSettings()));
+
+            collector.IntegrationRunning(IntegrationId);
+
+            var data = collector.GetData();
+            var integration = data.FirstOrDefault(x => x.Name == IntegrationName);
+            integration.Should().NotBeNull();
+            integration.AutoEnabled.Should().BeTrue();
+            integration.Enabled.Should().BeFalse();
```

Comment:
This is similar to my other comment about `IntegrationTelemetryCollector.GetData`, but I would expect `integration.Enabled` to equal true because we've decided to run our instrumentation code path. Here's my current interpretation of these values as they pertain to each integration, so if my interpretations are incorrect then that may be why I'm confused:
- AutoEnabled: At startup, we assign AutoEnabled for each integration because we know what the defaults are. Most integrations are run by default (assigned true) and some integrations may be in beta (assigned false), and this should be constant for the rest of the application lifetime.
- Enabled: At startup, each integration receives the result of `AutoEnabled`. If `AutoEnabled == false` but the configuration says to enable it, then we set `Enabled = true`. If at any time `Enabled == true` but the configuration says to disable it or we run into an instrumentation issue, then we set `Enabled = false`

Reply:
The approach I've used here was how we collectively decided to map our behaviour to the fixed flags  available via the telemetry API:

- `Enabled=true`: at least one span has been successfully emitted for that integration.
- `AutoEnabled`: the integration's code path has been successfully run. Note that this does _not_ mean it necessarily generated a span, just that the integration code path was entered.

The reasoning is that we can already infer the "default" values of the integrations from the tracer version. With the approach you describe above, we have no way of knowing if a) the integration code was ever executed (we would have `true`/`true` for a default-on integration that's never invoked) b) the integration was invoked but didn't generate a span (some required condition was not met for example)

However, there's also a (currently unused) flag, `Compatible`. Maybe we could use that flag to store the default value for the integration? That way we can preserve the most data?

Reply:
Thanks for the clarification, that makes sense to me and what you've implemented gives us more insights than the other approach I was thinking of.

Regarding the `Compatible` flag, I'm not sure it's worth using the flag to answer the default value question because we could use the flag in other ways to answer other important questions. For example, one way we could use it is to set that flag when the application has loaded a lower (or newer) version of an library that should receive automatic instrumentation but doesn't because of the incorrect version, which would then help us understand what unsupported versions are being run in the wild that we don't currently support.

Reply:
> I'm not sure it's worth using the flag to answer the default value question because we could use the flag in other ways to answer other important questions

Yeah, that's a good point, once we use it for that, we can't use it for anything.

> one way we could use it is to set that flag when the application has loaded a lower (or newer) version of an library that should receive automatic instrumentation but doesn't because of the incorrect version

_Theoretically_ yes, though I'm not sure how to actually implement that at this stage 😄 Something for a follow up PR maybe

> which would then help us understand what unsupported versions are being run in the wild that we don't currently support

After this we will collect the name and version of all loaded assemblies, so we can probably already infer that information🙂